### PR TITLE
[dataset] new io for code reuse for many speech tasks

### PR DESCRIPTION
**Pull request by Mddct**
https://github.com/wenet-e2e/wenet/issues/2152
- TODO
   -  [x] chain call
   -  [x] support raw dataset source
   -  [x] support tar shard dataset source
   -  [x] prefetch in synchronize  way
   -  [x] mapper Ignore error
   -  [x] sort data pipes
   -  [x] check shuffle deterministic
   -  [x] batching 
         - [x] static batch
         - [x] dynamic batch
    - [x] padding
    - [x] unit test 
        - [x] raw dataset
        - [x] tar dataset
        - [x] unit test for all related function in processor.py 
    - [x] reproduce exp 
       - [x]  raw
       - [x] shard

- next pr todo
   -  [ ] bucket batch
   -  [ ] multiprocessing prefetch for CPU bound operation
   -  [ ] cv support distributed and join 
   -  [ ] 修复断点处的种子

### CODE REVIEW

**Comment by Mddct** [comment]
chain works:
```python
from torch.utils.data import DataLoader
from wenet.dataset.datapipes import (WenetRawDatasetSource,
                                     WenetTarShardDatasetSource)
from wenet.dataset.processor_v2 import (compute_fbank,
                                        compute_log_mel_spectrogram,
                                        decode_wav, parse_json)

if __name__ == '__main__':

    dataset = WenetTarShardDatasetSource(
        'test/resources/dataset/data.shards.list')
    # dataset = dataset.map(decode_wav)
    # dataset = dataset.map(compute_fbank)
    # dataset = dataset.map(compute_log_mel_spectrogram)
    dataloader = DataLoader(dataset,
                            num_workers=1,
                            persistent_workers=True,
                            batch_size=None)
    for d in dataloader:
        print(d['file_name'], d['tar_file_name'], d['txt'], len(d['wav']))
    for d in dataloader:
        print(d['file_name'], d['tar_file_name'], d['txt'], len(d['wav']))
    for d in dataloader:
        print(d['file_name'], d['tar_file_name'], d['txt'], len(d['wav']))

    dataset = dataset.map(decode_wav).map(compute_fbank)
    dataloader = DataLoader(dataset,
                            num_workers=1,
                            persistent_workers=True,
                            batch_size=None)
    for d in dataloader:
        print(d['file_name'], d['tar_file_name'], d['txt'], len(d['wav']),
              d['feat'].size())

    dataset = WenetRawDatasetSource("test/resources/dataset/data.list",
                                    prefetch=10)
    dataset = dataset.map(parse_json)
    dataset = dataset.map(decode_wav)
    dataset = dataset.map(compute_log_mel_spectrogram)
    dataloader = DataLoader(dataset,
                            num_workers=2,
                            persistent_workers=True,
                            batch_size=None)

    for d in dataloader:
        print(d.keys())

    for d in dataloader:
        print(d.keys())

```

Screenshot 2024-01-23 at 16 34 49

**Comment by Mddct** [comment]
batch work
```python
    dataset = WenetRawDatasetSource("test/resources/dataset/data.list",
                                    prefetch=10)
    dataset = dataset.map(parse_json)
    dataset = dataset.map(decode_wav)
    dataset = dataset.map(compute_log_mel_spectrogram)

    def fake_labels(sample):
        sample['label'] = [1, 2, 3, 4]
        return sample

    dataset = dataset.map(fake_labels)
    dataset = dataset.batch(batch_size=2, wrapper_class=padding)

    for d in dataset:
        print(d)
```
Screenshot 2024-01-23 at 18 16 25

**Comment by Mddct** [comment]
dynamic batch work

Screenshot 2024-01-23 at 22 30 20

File: test/wenet/dataset/test_datapipes.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+from functools import partial
+import torch
+from torch.utils.data import datapipes
+
+from wenet.dataset.datapipes import SortDatPipes
```
Comment (xingchensong) [inline test/wenet/dataset/test_datapipes.py:5]: SortDataPipes

File: test/wenet/dataset/test_dataset.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import pytest
+import torch
+from wenet.dataset.datapipes import WenetTarShardDatasetSource
+from wenet.dataset.dataset_v2 import WenetRawDatasetSource
```
Comment (xingchensong) [inline test/wenet/dataset/test_dataset.py:4]: from wenet.dataset.datapipes import WenetRawDatasetSource

File: wenet/dataset/dataset_v2.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    dataloader = DataLoader(dataset,
+                            num_workers=1,
+                            persistent_workers=True,
+                            batch_size=None)
+    for d in dataloader:
+        print(d['file_name'], d['tar_file_name'], d['txt'], len(d['wav']))
+    for d in dataloader:
+        print(d['file_name'], d['tar_file_name'], d['txt'], len(d['wav']))
+    for d in dataloader:
+        print(d['file_name'], d['tar_file_name'], d['txt'], len(d['wav']))
```
Comment (xingchensong) [inline wenet/dataset/dataset_v2.py:30]: why repeat 3 times?

Comment (Mddct) [reply wenet/dataset/dataset_v2.py:30]: 目前只是个测试文件（测试是否开始新的迭代） 忽略 
后边会串起来 替换当前的dataset.py文件

File: wenet/dataset/datapipes.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+                    sample['process'].communicate()
+                sample['stream'].close()
+
+
+class WenetRawDatasetSource(IterDataPipe):
+
+    def __init__(self, filenames: str, prefetch: int = 500) -> None:
+        super().__init__()
+        self.dp = TextLineDataPipes(filenames).prefetch(
+            prefetch).sharding_filter()
```
Comment (xingchensong) [inline wenet/dataset/datapipes.py:275]: what does `sharding_filter()` do?

File: wenet/dataset/datapipes.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+                self._buffer.append(elem)
+            else:
+                if len(self._buffer) > 0:
+                    yield self._wrappr_class(self._buffer)
+                del self._buffer
+                self._buffer = [elem]
+        if len(self._buffer) > 0:
+            yield self._wrappr_class(self._buffer)
+        del self._buffer
+        self._buffer = []
```
[the remark this answers is not in the document]
Comment (Mddct) [reply wenet/dataset/datapipes.py:88]: yes

**Comment by Mddct** [comment]
https://github.com/pytorch/pytorch/blob/main/torch/utils/data/dataloader.py#L169

**Comment by Mddct** [comment]
测试多机： 数据集 aishell dev， 条数14326
```python
from torch.utils.data import DataLoader
from wenet.dataset.datapipes import WenetRawDatasetSource
from wenet.dataset.processor import decode_wav, parse_json
import os
import torch.distributed as dist
import torch

world_size = int(os.environ.get('WORLD_SIZE', 1))
local_rank = int(os.environ.get('LOCAL_RANK', 0))
rank = int(os.environ.get('RANK', 0))

dist.init_process_group('nccl')

data_list = '/mnt/afs/zhoudinghao/work/wenet3/wenet/examples/test/wenet_format.dev.list'
dataset = WenetRawDatasetSource(data_list)
dataset = dataset.map(parse_json)
dataset = dataset.map(decode_wav)

generator = torch.Generator()
generator.manual_seed(777)
dataloader = DataLoader(dataset,
                        batch_size=None,
                        pin_memory=False,
                        num_workers=10,
                        persistent_workers=True,
                        generator=generator,
                        prefetch_factor=100)

dist.barrier()

for d in dataset:
    print(d['sample_rate'])
```
输出条数
```bash
14326 test.py.log
```

**Comment by Mddct** [comment]
- 测试二： uneven data ， 模拟loss reduce sum, aishell dev 14326 条， 2卡， num worker 为10

则 14326 // （2*10） ！= 0
```

world_size = int(os.environ.get('WORLD_SIZE', 1))
local_rank = int(os.environ.get('LOCAL_RANK', 0))
rank = int(os.environ.get('RANK', 0))

dist.init_process_group('nccl')

data_list = '/mnt/afs/zhoudinghao/work/wenet3/wenet/examples/test/wenet_format.dev.list'
dataset = WenetRawDatasetSource(data_list)
dataset = dataset.map(parse_json)
dataset = dataset.map(decode_wav)

generator = torch.Generator()
generator.manual_seed(777)
dataloader = DataLoader(dataset,
                        batch_size=None,
                        pin_memory=False,
                        num_workers=10,
                        persistent_workers=True,
                        generator=generator,
                        prefetch_factor=100)
device = torch.device('cuda')
dist.barrier()

for d in dataset:
    wav = d['wav']
    wav = wav.to(device)
    value = wav[0][0]
    dist.all_reduce(value, op=torch.distributed.ReduceOp.SUM)
    print(value)

```

```bash
14326 test.py.log
````

行数对应， 测试通过， 后续需要修改CV 的loss 为all reduce mean的结果 （TODO）

**Comment by Mddct** [comment]
CV的处理方式和wenet 保持一致（uneven data）， 后续pr优化这部分

**Comment by Mddct** [comment]
new io 和old io loss下将曲线几乎一致
Screenshot 2024-01-27 at 10 22 55

**Comment by Mddct** [comment]
raw 解码复现, 比readme 好了一丢丢


Screenshot 2024-01-27 at 17 45 18

**Comment by Mddct** [comment]
aishell train_transformer.yaml 训练时间对比

| Syntax      |old io | new io| 
| ----------- | ----------- | ----------- | 
|    耗时        | 20h 5min.  | 17h20min  |

**Comment by Mddct** [comment]
shard 模式下 new io 和old io loss下将曲线几乎一致
Screenshot 2024-01-28 at 09 48 25

File: wenet/dataset/datapipes.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        super().__init__()
+        _dp = datapipes.iter.FileLister(filenames)
+        _dp = datapipes.iter.FileOpener(_dp, mode=mode)
+        self.dp = _dp
+
+    def __iter__(self):
+        for fname, stream in self.dp:
+            for line in stream:
+                line = line.strip('\n')
+                yield {"file_name": fname, "line": line}
```
Comment (kobenaxie) [inline wenet/dataset/datapipes.py:179]: What's the usage of `file_name` here ?

Comment (Mddct) [reply wenet/dataset/datapipes.py:179]: we support multiple index files:
for example:
raw mode: a.txt, b.txt ...as input，
tar shard mode: a.shard.list b.shard.list ...

file_name ： This field can be used for traceability

Comment (kobenaxie) [reply wenet/dataset/datapipes.py:179]: Ok, thanks, get it.

**Comment by Mddct** [comment]
shard 模式结果复现

Screenshot 2024-01-28 at 18 30 26

File: wenet/dataset/dataset.py [hunk 7]
```diff
@@ -133,70 +20,87 @@ def Dataset(data_type,
...
+    if batch_type == 'static':
+        assert 'batch_size' in batch_conf
+        batch_size = batch_conf.get('batch_size', 16)
+        dataset = dataset.batch(batch_size, wrapper_class=processor.padding)
+    else:
+        max_frames_in_batch = batch_conf.get('max_frames_in_batch', 12000)
+        dataset = dataset.dynamic_batch(
+            processor.DynamicBatchWindow(max_frames_in_batch),
+            wrapper_class=processor.padding,
+        )
```
Comment (xingchensong) [inline wenet/dataset/dataset.py:104]: 为何有的需要partial（tokenize，filter，resample，...），有的不需要呢（decode_wav）？

File: wenet/dataset/datapipes.py [hunk 8]
```diff
@@ -0,0 +1,330 @@
+import collections
```
Comment (xingchensong) [inline wenet/dataset/datapipes.py:1]: 加一下你的copyright

File: wenet/utils/train_utils.py [hunk 9]
```diff
@@ -261,12 +263,14 @@ def init_dataset_and_dataloader(args, configs, tokenizer):
                                    pin_memory=args.pin_memory,
                                    num_workers=args.num_workers,
                                    persistent_workers=True,
+                                   generator=generator,
```
Comment (xingchensong) [inline wenet/utils/train_utils.py:266]: 接上一个断点继续训练的问题，generator.manual_seed(epoch)更符合原来的含义一些，另外generator.manual_seed(epoch+step)可能会更好（有时是按照step模式存的ckpt）。

epoch和step信息应该可以从configs里拿到，initdataset挪到initmodel后面
image

此时可以通过如下方式获取step，epoch同理
image

Comment (xingchensong) [reply wenet/utils/train_utils.py:266]: 不过这个问题可以下一个PR修复，这里纪录一下

File: wenet/dataset/dataset.py (hunk 7 above)
Comment (Mddct) [reply wenet/dataset/dataset.py:104]: processor 对应函数除了same，有其他参数的（各种conf） 需要partial绑定一下

decode wav只有一个sample参数

Comment (xingchensong) [reply wenet/dataset/dataset.py:104]: get

File: wenet/bin/train.py [hunk 10]
```diff
@@ -127,7 +127,6 @@ def main():
     configs.pop("init_infos", None)
     final_epoch = None
     for epoch in range(start_epoch, configs.get('max_epoch', 100)):
-        train_dataset.set_epoch(epoch)
```
[the remark this answers is not in the document]
Comment (Mddct) [reply wenet/bin/train.py:130]: https://github.com/pytorch/pytorch/blob/main/torch/utils/data/dataloader.py#L570-L575

后续pr修复， 端点处的shuffle （step： epoch+step）

**Comment by Mddct** [comment]
稍等 ignore error 有些bug

**Review by xingchensong — APPROVED** [review]

**Review by robin1001 — APPROVED** [review]

**Comment by xingchensong** [comment]
great job！

**Comment by robin1001** [comment]
Big job!
